### DISCUSSION TITLE
Authoring plugins to automatically add imports

### DISCUSSION THREAD

**Discussion by JoviDeCroock** [swc-project/swc#1378] [Q&A]
Hey all,

Loving the swc-project, was wondering whether or not it's possible to create plugins that add imports, for instance when someone uses JSX we can automatically add that import, ...

I made an example:

<details>

```
class JSXImportAppender extends Visitor.default {
	visitProgram(e) {
		const preactImport = e.body.find(d => d.type === 'ImportDeclaration' && d.source && d.source.value === 'preact');
		if (!preactImport) {
			return {
				...e,
				body: [
					{
						type: 'ImportDeclaration',
						specifiers: [
							{
								type: 'ImportSpecifier',
								local: {
									type: 'Identifier',
									value: 'h'
								}
							},
							{
								type: 'ImportSpecifier',
								local: {
									type: 'Identifier',
									value: 'Fragment'
								}
							}
						],
						source: {
							type: 'StringLiteral',
							value: 'preact'
						}
					},
					...e.body
				]
			};
		}

		if (!preactImport.specifiers.find(x => x.local.value === 'h')) {
			preactImport.specifiers.push({
				type: 'ImportSpecifier',
				local: {
					type: 'Identifier',
					value: 'h'
				}
			});
		}

		if (!preactImport.specifiers.find(x => x.local.value === 'Fragment')) {
			preactImport.specifiers.push({
				type: 'ImportSpecifier',
				local: {
					type: 'Identifier',
					value: 'Fragment'
				}
			});
		}

		return e;
	}
}
```

</details>

However this seems to always fail with the message that we are missing the span attributes

```
thread '<unnamed>' panicked at 'failed to deserialize Program: Error("missing field `span`", line: 0, column: 0)', native\src\transform.rs:42:50
```

Is there a way to automatically generate these or is this purposely unsupported.

Another issue I'm running into (Disclaimer using this through rollup) is that when doing multiple files concurrently the plugin doesn't seem to fire when using it like this: `plugin: m => new JSXImportAppender().visitProgram(m),`

Relevant PR: https://github.com/preactjs/wmr/pull/321/files#r571402713

**Comment by kapouer** [comment]
Hi i was playing with an swc plugin to rewrite imports paths.

> Is there a way to automatically generate these or is this purposely unsupported.

Instead of passing a handmade tree, i'd try to get a new tree out of swc.parse() and pluck it in the right place.

> Another issue I'm running into (Disclaimer using this through rollup) is that when doing multiple files concurrently the plugin doesn't seem to fire

That's related to https://github.com/swc-project/swc/commit/a0898e8ce311d5838c52a77dd5e47974e21016ac
